### PR TITLE
Implement providing drug frequencies label depending on the country in `DrugSearchScreen`

### PR DESCRIPTION
**Pull request by janhavisinghh**
https://app.shortcut.com/simpledotorg/story/5043/implement-providing-drug-frequencies-label-depending-on-the-country-in-drugsearchscreen

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/drugs/search/DrugsSearchScreen.kt [hunk 1]
```diff
@@ -128,7 +131,7 @@ class DrugsSearchScreen : BaseScreen<
   override fun setDrugSearchResults(searchResults: PagingData<Drug>, drugFrequencyToFrequencyChoiceItemMap: Map<DrugFrequency?, DrugFrequencyChoiceItem>) {
     val searchQuery = searchQueryEditText.text?.toString().orEmpty()
     drugSearchResultsList.scrollToPosition(0)
-    adapter.submitData(lifecycle, DrugSearchListItem.from(searchResults, searchQuery))
+    adapter.submitData(lifecycle, DrugSearchListItem.from(searchResults, searchQuery, drugFrequencyToFrequencyChoiceItemMap))
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/search/DrugsSearchScreen.kt:134]: Just an FYI, this should be done in the next commit.

File: app/src/main/java/org/simple/clinic/drugs/search/DrugSearchModel.kt [hunk 2]
```diff
@@ -2,16 +2,19 @@ package org.simple.clinic.drugs.search
 
 import android.os.Parcelable
 import kotlinx.parcelize.Parcelize
+import org.simple.clinic.drugs.selection.custom.drugfrequency.country.DrugFrequencyChoiceItem
 
 @Parcelize
 data class DrugSearchModel(
-    val searchQuery: String
+    val searchQuery: String,
+    val drugFrequencyToFrequencyChoiceItemMap: Map<DrugFrequency?, DrugFrequencyChoiceItem>?
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/search/DrugSearchModel.kt:10]: If we are using this in the UI layer only, can't we just inject it directly to the adapter or screen? WDYT? 🤔

File: app/src/main/java/org/simple/clinic/drugs/search/DrugSearchUpdate.kt [hunk 3]
```diff
@@ -17,10 +16,19 @@ class DrugSearchUpdate : Update<DrugSearchModel, DrugSearchEvent, DrugSearchEffe
...
-      is DrugFrequencyChoiceItemsLoaded -> noChange()
+      is DrugFrequencyChoiceItemsLoaded -> drugFrequencyChoiceItemsLoaded(model, event)
     }
   }
 
+  private fun drugFrequencyChoiceItemsLoaded(
+      model: DrugSearchModel,
+      event: DrugFrequencyChoiceItemsLoaded
+  ): Next<DrugSearchModel, DrugSearchEffect> {
+    val drugFrequencyToFrequencyChoiceItemMap = event.drugFrequencyChoiceItems.items.associateBy({ it.drugFrequency }, { it })
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/search/DrugSearchUpdate.kt:27]: I see we are converting the list to a map for the drug choice items, can't we just provide it as a map directly?

**Review by msasikanth — CHANGES_REQUESTED** [review]
Looks good overall, I have left one question related to the approach. Also, I feel `DrugChoiceItems` might be a good name for this, we might have to consider changing it in future. For example, the name indicates we are using it for choice items, but we are using it for the conversion of drug frequency labels.

File: app/src/main/java/org/simple/clinic/drugs/search/DrugSearchUpdate.kt (hunk 3 above)
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/drugs/search/DrugSearchUpdate.kt:27]: Umm.. We are mapping it as DrugFrequency or MedicineFrequency to DrugFrequencyChoiceItem. Although you are right that approach sounds better and will remove a lot of code. We can merge this in for now. I will make that change in all the places and remove the unnecessary code.

Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/drugs/search/DrugSearchUpdate.kt:27]: Yeah, sounds good. I am not saying we need to do this change here. I was just checking to see if there was any particular reason for going with this approach. We can also combine `DrugFrequency` & `MedicineFrequency` into one thing, right? Technically, I don't think there is any difference between them 🤔

Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/drugs/search/DrugSearchUpdate.kt:27]: Yeah, we can take the whole DrugFrequency implementation out in a separate package and use DrugFrequency everywhere. Maintaining two different data classes is not necessary at all here.

File: app/src/main/java/org/simple/clinic/drugs/search/DrugsSearchScreen.kt [hunk 4]
```diff
@@ -49,6 +52,11 @@ class DrugsSearchScreen : BaseScreen<
   @Inject
   lateinit var router: Router
 
+  @Inject
+  lateinit var drugFrequencyFactory: DrugFrequencyFactory
+
+  private var drugFrequencyToFrequencyChoiceItemMap: Map<DrugFrequency?, DrugFrequencyChoiceItem>? = null
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/search/DrugsSearchScreen.kt:58]: Is there a reason why it has to be a mutable property? and also can't we just have this inside the `onViewCreated` as well?

File: app/src/main/java/org/simple/clinic/drugs/search/DrugsSearchScreen.kt [hunk 5]
```diff
@@ -105,6 +113,9 @@ class DrugsSearchScreen : BaseScreen<
   override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
     super.onViewCreated(view, savedInstanceState)
 
+    val drugFrequencyChoiceItems = DrugFrequencyChoiceItems(drugFrequencyFactory.provideFields())
+    drugFrequencyToFrequencyChoiceItemMap = drugFrequencyChoiceItems.items.associateBy({ it.drugFrequency }, { it })
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/search/DrugsSearchScreen.kt:117]: Can't we just use this instead? 🤔 . We can even do this inside the adapter as well and just pass in the factory/items in the screen.

```suggestion
val drugFrequencyToFrequencyChoiceItemMap = drugFrequencyFactory
    .provideFields()
    .associateBy({ it.drugFrequency }, { it })
```

File: app/src/main/java/org/simple/clinic/drugs/search/DrugSearchListItem.kt [hunk 6]
```diff
@@ -21,9 +22,9 @@ sealed class DrugSearchListItem : PagingItemAdapter.Item<Event> {
 
   companion object {
 
-    fun from(searchResults: PagingData<Drug>, searchQuery: String): PagingData<DrugSearchListItem> {
+    fun from(searchResults: PagingData<Drug>, searchQuery: String, drugFrequencyToFrequencyChoiceItemMap: Map<DrugFrequency?, DrugFrequencyChoiceItem>): PagingData<DrugSearchListItem> {
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/search/DrugSearchListItem.kt:25]: Can we put the arguments in separate lines, please? It crosses our character limit in a single line.

File: app/src/main/java/org/simple/clinic/drugs/search/DrugSearchListItem.kt [hunk 7]
```diff
@@ -49,7 +50,7 @@ sealed class DrugSearchListItem : PagingItemAdapter.Item<Event> {
     }
   }
 
-  data class DrugSearchResult(val drug: Drug) : DrugSearchListItem() {
+  data class DrugSearchResult(val drug: Drug, val drugFrequencyToFrequencyChoiceItemMap: Map<DrugFrequency?, DrugFrequencyChoiceItem>) : DrugSearchListItem() {
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/search/DrugSearchListItem.kt:53]: Same comment as above

File: app/src/main/java/org/simple/clinic/drugs/search/DrugsSearchScreen.kt [hunk 8]
```diff
@@ -138,7 +138,7 @@ class DrugsSearchScreen : BaseScreen<
   override fun setDrugSearchResults(searchResults: PagingData<Drug>) {
     val searchQuery = searchQueryEditText.text?.toString().orEmpty()
     drugSearchResultsList.scrollToPosition(0)
-    adapter.submitData(lifecycle, DrugSearchListItem.from(searchResults, searchQuery))
+    adapter.submitData(lifecycle, DrugSearchListItem.from(searchResults, searchQuery, drugFrequencyToFrequencyChoiceItemMap!!))
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/search/DrugsSearchScreen.kt:141]: So yeah, if we follow the above suggestion where we create this inside the adapter. We can avoid creating a nullable property in the screen.

**Review by msasikanth — CHANGES_REQUESTED** [review]
LGTM. Left couple of comments PTAL

**Review by msasikanth — APPROVED** [review]

**Pull request merged**